Update to Eslint 9

ChristianPraiss opened the pull request:
> This is a draft and could probably be improved by stripping out more stuff.
> A preliminary review could not hurt, even though it is still in draft.
> All version will get a major increase so people will be aware of breaking changes with ESLint 9

### TODOS:

- [x] Test with 2 existing projects (remex, aio)

ChristianPraiss committed:
chore: upgrade packages

ChristianPraiss committed:
chore: full eslint upgrade

ChristianPraiss committed:
chore: add docs and migration plan

ChristianPraiss committed:
chore: initial migration

ChristianPraiss committed:
align node types with lts release

ChristianPraiss commented on packages/eslint-config-convidera/package.json:
```diff
@@ -1,21 +1,25 @@
   "publishConfig": {
     "access": "public"
+  },
+  "devDependencies": {
```
just realized i could remove eslint here again

deathmaz commented:
didn't check everything just briefly skimmed through, but already few questions:

- in https://github.com/convidera/frontend-kit/pull/13/commits/a5243f42b7b44b871accbcd2fcc324a73b5ab5ad a publish was made, but i don't see the updated versions [here](https://www.npmjs.com/settings/convidera-team/packages?page=0&perPage=10), what is the meaning of this commit? Was the publish reverted?
- why the versions in the mentioned commit are already "final" and not `alpha` or `beta` if the work on this PR still not finished? 
- there's no need to publish to npmjs if you need to test something locally, you can use `yarn pack` to create needed artifacts and then install those via `yarn install`
- if you want to test not only locally please release `alpha` or `beta` version and not `final`

ChristianPraiss commented:
Hey, I did not actually publish anything yet, just let it run through once to check if all works with the newest version, thus the versions also have not alpha beta or anything applied to them. Would revert/rerun the publish again once it is ready

ChristianPraiss commented:
Hey @deathmaz @kalievsky, will you guys have some time to check this soon?

deathmaz commented:
Hi @ChristianPraiss cannot promise when exactly but will check at some point, want to also test it in one of my projects. Have you tried it on real projects on your side?

deathmaz commented on packages/eslint-config-convidera/index.js:
```diff
@@ -1,86 +1,128 @@
+        module: 'readonly',
+        require: 'readonly',
+        exports: 'readonly',
       },
```
in the [docs](https://eslint.org/docs/latest/use/configure/language-options#predefined-global-variables) they suggest to use `globals` package to enable all globals for the specific environment. Is there a reason it's better to do this by hand and only define ones in this PR?

deathmaz commented on packages/eslint-config-vue3-convidera/index.js:
```diff
@@ -1,27 +1,47 @@
+        defineProps: 'readonly',
+        defineEmits: 'readonly',
+        defineExpose: 'readonly',
+        withDefaults: 'readonly',
```
there are also:
- defineSlots
- defineModel
- defineOptions

deathmaz commented on packages/utils/package.json:
```diff
@@ -20,21 +21,21 @@
     "lint": "yarn lint:js",
     "lint:fix": "yarn lint:js:fix",
-    "lint:js": "eslint --ext .js,.vue,.ts --ignore-path .gitignore .",
+    "lint:js": "eslint --ext .js,.vue,.ts .",
```
is it still needed to have `--ext` flag and having `files` in eslint.config ?

deathmaz commented on packages/utils/eslint.config.js:
```diff
@@ -0,0 +1,28 @@
+
+  // Project-specific configuration
+  {
+    files: [
+      '**/*.ts',
+      '**/*.tsx',
+      '**/*.js',
+      '**/*.jsx',
+    ],
```
in package json the command to run eslint is `eslint --ext .js,.vue,.ts .` Is it still needed to specify `--ext` flag and having `files` configured?

deathmaz replied:
probably also worth including `.mts` and `.mjs` files

deathmaz commented on packages/vue-form-validator/package.json:
```diff
@@ -25,25 +26,25 @@
     "lint": "yarn lint:js",
     "lint:fix": "yarn lint:js:fix",
-    "lint:js": "eslint --ext .js,.vue,.ts --ignore-path .gitignore .",
+    "lint:js": "eslint --ext .js,.vue,.ts .",
```
is it needed to have `--ext` specified and having `files` configured in estling.config?

deathmaz commented on packages/vue-form-validator/eslint.config.js:
```diff
@@ -0,0 +1,29 @@
+  // Project-specific configuration
+  {
+    files: [
+      '**/*.vue',
+      '**/*.ts',
+      '**/*.tsx',
+      '**/*.js',
+      '**/*.jsx',
```
also inlclude `.mjs` and `.mts`?

deathmaz commented on packages/vue-form-validator/src/form-validator/utils/is-valid-url.ts:
```diff
@@ -1,8 +1,10 @@
+/* global URL */
```
imo it's better to have it defined in eslint.config or as suggested before maybe it's worth using `globals` package in the base eslint config to cover all possible cases

deathmaz reviewed (commented):
- Stylelint also was bumped few major version but i cannot verify if it still works fine as I'm not using it any more
- some of the packages can be updated again since this PR was created
- not yet tested in real project

ChristianPraiss committed:
chore: update ESLint configurations and dependencies

- Added `globals` package to ESLint config for better global variable management.
- Updated dependencies in `eslint-config-convidera`, `eslint-config-ts-convidera`, and `eslint-config-vue3-convidera` to the latest versions.
- Adjusted ESLint rules and configurations for improved code quality and consistency.
- Removed deprecated or unnecessary configurations in ESLint setup.

This update enhances the linting process and ensures compatibility with the latest ESLint features.

ChristianPraiss committed:
chore: enhance ESLint configurations for TypeScript and JavaScript support

- Added support for `.mts` and `.mjs` file extensions in ESLint configurations for both utils and vue-form-validator packages.
- Simplified the linting command in package.json files to use default extensions.

This update improves linting capabilities for modern JavaScript and TypeScript file formats.

ChristianPraiss committed:
chore: extend ESLint configuration for Vue 3

- Added support for `defineOptions`, `defineSlots`, and `defineModel` as readonly properties in the ESLint configuration for Vue 3.

This update enhances linting capabilities for Vue 3 features, ensuring better code quality and consistency.

ChristianPraiss commented:
Soo, I updated it and gave it a round of testing. It's a slight bit stricter overall (tested on https://github.com/convidera/partnerlink), reporting some no-unused-vars and no-explicit-any warnings, but i guess those are very valid rules anyway. @jhoefker @deathmaz.

deathmaz commented:
@ChristianPraiss i tried it on my project and in some places eslint reported errors where it shouldn't:

<details>

<img width="872" height="154" alt="image" src="https://github.com/user-attachments/assets/5a006850-df63-4804-a11a-9041f78d261c" />

</details>

in all `defineEmits` declared as:

```typescript
const emit = defineEmits<{(e: 'close'): void,
  (e: 'success', payload: AppDataTaskApplicationFeedFeedItemResource): void
}>();

```

things like `e` and `payload` reported as unused vars. There's no error if emits are defined this way:

```typescript
const emit = defineEmits<{
  close: [],
  success: [payload: AppDataTaskApplicationFeedFeedItemResource]
}>();
```

Because those are not errors and the syntax is a valid vue3 syntax, imo it's better to not report those things as errors if possible.

The rest of the reported issues i got mostly about no-explicit-any, which if needed could be changed on per project basis and i'm fine with it

ChristianPraiss commented:
Hey, thx for testing it out! Will check what can be done about the rule, will see if there is a vue-specific version or if the rule params have an option to fix this.

deathmaz commented:
@ChristianPraiss will you finalise the PR?
